### ISSUE TITLE
New lint: Have a semicolon on the last block statement if it returns nothing

### ISSUE THREAD

**Issue by xFrednet** [rust-lang/rust-clippy#6467]
### What it does

Validates that the last block statement ends with a semicolon if it returns nothing.

### Categories (optional)

- Kind: `clippy::style`

### Advantages
* It leads to consistent block formatting.
* Extending the block with new code doesn't require a change in previous last line
    <details>
    <summary>Example:</summary>

    ```rust
    // Valid code
    {
        println!("Hello")
    }

    // Extending it would cause a change to the old `println!("Hello")` line
    {
        println!("Hello"); // <-- here
        println!("World")
    } 
    ```
    </details>
* It makes block linting easier for us
    * This is of course not the main reason but this causes some edge cases in a lint I'm currently working on.

### Drawbacks

None.

### Example

```rust
{
    println!("Hello world")
}
```

Could be written as:

```rust
{
    println!("Hello world");
}
```

**Comment by xFrednet** [comment]
I believe that this could be a good first issue. I'm not sure in which file it would be implemented by you could roughly do the following to implement this lint:

1. Filter out any [`rustc_hir::Block`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.Block.html) from the ast. (Maybe with [`check_block`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/trait.LateLintPass.html#method.check_block))
2. Check if the `expr` field is `Some`. (This means it has a statement without a semicolon at the end)
3. Get the type of `expr` from a [`LateContext`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/context/struct.LateContext.html) instance
4. Check if the type is `()` with [`.is_unit()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_unit) on the type.

@rustbot label +"good first issue"

**Comment by flip1995** [comment]
Regarding the lint level: This should probably be pedantic, since it doesn't really matter. Great explanation how to implement this @xFrednet!

**Comment by camsteffen** [comment]
I can't get an example for this lint to compile. Am I missing something? ([playground])(https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0ca1c9906aaccd73fa77d29eaba0ac8c)

**Comment by xFrednet** [comment]
You're right @camsteffen the expression has to return the unit type `()` I've missed that. Thank you! I've updated the description. [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d3ce8444ce2195d5a84acd1fe669d35)

**Comment by camsteffen** [comment]
Cool. I guess the difference is that `let x = ...` is not an expression - it's a statement.

**Comment by camsteffen** [comment]
Two questions.

1. What if the expression does not return `()`, but the value is unused?
    ```rust
    fn test() {
        let mut set = HashSet::new();
        {
            set.insert("hi")
        }
    }
    ```
2. Is this a job for rustfmt?

**Comment by flip1995** [comment]
1. Your example will produce a `mismatched type` error. You can only leave out the semicolon if the last expression in the block returns ``()``
2. No idea if rustfmt does type analysis.

**Comment by camsteffen** [comment]
Ah nevermind. I thought this might not require type checking.

**Comment by 1c3t3a** [comment]
I've started to implement this [on my own fork](https://github.com/1c3t3a/rust-clippy/blob/1c3t3a-issue-6467/clippy_lints/src/semicolon_if_nothing_returned.rs) but ran into an issue while testing. I first tested the lint on the `println!()` macro where it worked alright. Looking at `assert_eq!()`, my lint fired even with a `;` at the end of the line. As I assumed the [`assert_eq!()`](https://doc.rust-lang.org/src/core/macros/mod.rs.html#42-75) macro implementation doesn't end with a semicolon, but the [`println!()`](https://doc.rust-lang.org/src/std/macros.rs.html#60-62) macro does. I know this might be a beginner question, but is there a way to check for a semicolon after the macro invocation? I've already experimented with `in_macro` but it didn't quite help.

**Comment by camsteffen** [comment]
Is the lint firing on the block that is part of the macro expansion?

```rust
macro_rules! assert_eq {
    ($left:expr, $right:expr) => ({ // <- don't lint on the block that starts here
```

You could do `dbg!(block.span)` and look at the line/column numbers to find out.

**Comment by 1c3t3a** [comment]
> Is the lint firing on the block that is part of the macro expansion?
> 
> ```rust
> macro_rules! assert_eq {
>     ($left:expr, $right:expr) => ({ // <- don't lint on the block that starts here
> ```
> 
> You could do `dbg!(block.span)` and look at the line/column numbers to find out.

Actually yes, it's exactly that line.

**Comment by camsteffen** [comment]
If you need more help, head over to [zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).
